src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -25,22 +25,185 @@
+
+                std::string s(iter_ctx.filename.begin(), iter_ctx.filename.end());
+                auto icf = iter_ctx.ctx.get_hooks().preprocessOptions.includeFinder; // TODO it's null?
+                //std::optional<std::string> st = icf.getIncludeStandard(p, s);
+                //unfortunately have to call get_resource_runtime here again
+                auto builtin_pair = nbl::builtin::get_resource_runtime(s);
+                if (builtin_pair.first) //not a nullptr
+                {
+                    iter_ctx.instring = std::string(builtin_pair.first, builtin_pair.first+builtin_pair.second);
+                }
+                else {
+                    //TODO use nbl filesystem
+                    // read in the file
+                    boost::filesystem::ifstream instream(iter_ctx.filename.c_str());
+                    if (!instream.is_open()) {
+                        BOOST_WAVE_THROW_CTX(iter_ctx.ctx, boost::wave::preprocess_exception,
+                            bad_include_file, iter_ctx.filename.c_str(), act_pos);
+                        return;
+                    }
+                    instream.unsetf(std::ios::skipws);
+
+                    iter_ctx.instring.assign(
+                        std::istreambuf_iterator<char>(instream.rdbuf()),
+                        std::istreambuf_iterator<char>());
+                }
+                iter_ctx.first = iterator_type(
+                    iter_ctx.instring.begin(), iter_ctx.instring.end(),
+                    PositionT(iter_ctx.filename), language);
+                iter_ctx.last = iterator_type();
+            }
+
+        private:
+            std::string instring; //todo change to IReferenceCounted file
+        };
+    };
+
+    struct custom_preprocessing_hooks : public boost::wave::context_policies::default_preprocessing_hooks{
+
+        custom_preprocessing_hooks(const nbl::asset::IShaderCompiler::SPreprocessorOptions& _preprocessOptions, std::vector<std::pair<uint32_t, std::string>>&& _lineOffsetStack) : preprocessOptions(_preprocessOptions), lineOffsetStack(_lineOffsetStack) {}
+
+        nbl::asset::IShaderCompiler::SPreprocessorOptions preprocessOptions;
+        std::vector<std::pair<uint32_t, std::string>> lineOffsetStack;  //TODO no longer needed
+
+
+        template <typename ContextT>
+        bool
+            locate_include_file(ContextT& ctx, std::string& file_path,
+                bool is_system, char const* current_name, std::string& dir_path,
+                std::string& native_name) {
+
+            //on builtin return true
+            //default returns false if file does not exist
+            if (nbl::builtin::hasPathPrefix(file_path))
+            {
+                file_path = file_path.substr(nbl::builtin::pathPrefix.size() + 1);
+                return true;
+            }
+            if (!ctx.find_include_file(file_path, dir_path, is_system, current_name))
+                return false;   // could not locate file
```

Comment:
do you need to call it?

I'd call the old code here
```cpp

    std::optional<std::string> result;
    if (isRelative)
        result = inclFinder->getIncludeRelative(relDir, requestedSource);
    else //shaderc_include_type_standard
        result = inclFinder->getIncludeStandard(relDir, requestedSource);
```

btw `isRelative == !is_system` is the mapping of old stuff